azurerm/internal/services/compute/linux_virtual_machine_resource.go

```diff
@@ -828,6 +852,55 @@ func resourceLinuxVirtualMachineUpdate(d *schema.ResourceData, meta interface{})
 		}
 	}
 
+	// list of disks for expand operations
+	dataDisksForGrow := make([]map[string]interface{}, 0)
+	// list of disks with encryption set changes
+	dataDisksForEncrypt := make([]map[string]interface{}, 0)
+	if features.VMDataDiskBeta() && d.HasChange("data_disks") {
+		shouldUpdate = true
+
+		oldRaw, newRaw := d.GetChange("data_disks.0.local")
+		oldDisks := oldRaw.(*schema.Set).List()
+		newDisks := newRaw.(*schema.Set).List()
+		for _, o := range oldDisks {
+			oldDisk := o.(map[string]interface{})
+			if oldDiskName, ok := oldDisk["name"]; ok {
+				for _, n := range newDisks {
+					newDisk := n.(map[string]interface{})
+					if newDiskName, ok := newDisk["name"]; ok && oldDiskName.(string) == newDiskName.(string) {
+						if newDisk["disk_size_gb"].(int) < oldDisk["disk_size_gb"].(int) {
+							return fmt.Errorf("new disk size cannot be smaller than existing for %q, in Virtual Machine %q (resource group %q)", oldDisk["name"], id.Name, id.ResourceGroup)
+						} else if newDisk["disk_size_gb"].(int) > oldDisk["disk_size_gb"].(int) {
+							shouldShutDown = true
+							shouldDeallocate = true
+							dataDisksForGrow = append(dataDisksForGrow, newDisk)
+						}
+						if newDisk["disk_encryption_set_id"].(string) != oldDisk["disk_encryption_set_id"].(string) {
+							dataDisksForEncrypt = append(dataDisksForEncrypt, newDisk)
+						}
+					}
+				}
+			}
+		}
+		// EncryptionSet changes for "existing" disks
+		oldExistingRaw, newExistingRaw := d.GetChange("data_disks.0.existing")
+		oldExisting := oldExistingRaw.(*schema.Set).List()
+		newExisting := newExistingRaw.(*schema.Set).List()
+		for _, o := range oldExisting {
+			oldDisk := o.(map[string]interface{})
+			if oldDiskID, ok := oldDisk["managed_disk_id"]; ok {
+				for _, n := range newExisting {
+					newDisk := n.(map[string]interface{})
+					if newDiskID, ok := newDisk["managed_disk_id"]; ok && oldDiskID.(string) == newDiskID.(string) {
+						if newDisk["disk_encryption_set_id"].(string) != oldDisk["disk_encryption_set_id"].(string) {
+							dataDisksForEncrypt = append(dataDisksForEncrypt, newDisk)
+						}
+					}
+				}
+			}
+		}
+	}
```

Comment:
(same here) should we pull this out into a common helper method/type here? seems worthwhile to expose a method to resize the disks / encryption update which could potentially be done in parallel in the future - and shared across Linux & Windows in the interim?